styx-common/src/main/java/com/spotify/styx/storage/DatastoreStorage.java

```diff
@@ -761,20 +763,26 @@ void deleteResource(String id) throws IOException {
   }
 
   private void deleteShardsForCounter(String counterId) throws IOException {
-    final List<Key> shards = new ArrayList<>();
+    final List<Shard> shards = new ArrayList<>();
     datastore.query(EntityQuery.newEntityQueryBuilder()
         .setKind(KIND_COUNTER_SHARD)
         .setFilter(PropertyFilter.eq(PROPERTY_COUNTER_ID, counterId))
-        .build(), entity -> shards.add(entity.getKey()));
+        .build(), this::entityToShard);
```

Comment:
This feels stupid converting back and force, in order to make `StorageTransaction` look cleaner (not involving `Key` in the interface).

Reply:
Not valid comment.